Support yaml streams when decoding

chancez opened the pull request:
A port of my PR at the upstream repo to support yaml streams with the unmarshal methods: https://github.com/ghodss/yaml/pull/47

chancez commented:
/assign @smarterclayton

neolit123 commented:
/cc

dims commented:
@chancez do you mind throwing this change at the vendored copy of kubernetes/kubernetes and see what breaks if any? (just a WIP with a patched yaml.go in main kubernetes repo)

chancez commented:
@dims as a PR? sure

dims commented:
@chancez yes! so we can see if the CI catches any issues

chancez commented:
https://github.com/kubernetes/kubernetes/pull/73227 is up

chancez commented:
Got some failures in kube/kube...looking into it now.

chancez commented:
Ok, I found a way to reproduce one error (the panic) from kube/kube and it's related to unmarshalling a yaml document with the content `null`. Currently adding a test for it and debugging. It's reflection related.

dims commented:
very cool! thanks @chancez

chancez commented:
Okay, I fixed the panic I saw and added some tests for it. I'm looking into the other failures in kube/kube. I believe it's likely related to kube/kube having it's own logic/wrappers to handle streaming, but that's just a guess.

chancez commented:
Ah, how interesting, at least one kube/kube failure is that my code produces an `[]` when given `""` as the input. Before my change, it outputs `null`. The json unmarshaller behavior seems similar so that seems correct.

chancez commented:
Okay, I believe I fixed the kube/kube bazel test failures. I just pushed up with the updates I made in to this PR and it's re-running tests.

chancez commented:
Kube/kube is now only failing on godep and kops.

dims commented:
@chancez  looks like you worked through some thorny stuff which would have bitten us. nice! thanks.

dims commented:
@deads2k @sttts @lavalamp one of you wanna take a look? thanks!

chancez commented on yaml.go:
```diff
@@ -96,6 +96,62 @@ func JSONToYAML(j []byte) ([]byte, error) {
+// for decoding yaml streams.
+func internalYamlUnmarshalWrapper(strict bool, b []byte, out interface{}) error {
+	if len(bytes.TrimSpace(b)) == 0 {
+		out = nil
```
I realized this has no effect and I still need to use reflect here and down on 144 too. Sending a new commit for that.

chancez commented:
Squashed and made a minor fix to correctly set `out` using reflection when the input value is empty/null. I'm looking into benchmarking now. Do we want these to be just regular go test benchmarks? Trying to figure out what a good dataset would even be.

neolit123 commented on yaml.go:
```diff
@@ -96,6 +96,61 @@ func JSONToYAML(j []byte) ([]byte, error) {
+		outs = append(outs, tmp)
+	}
+
+	if len(outs) == 0 {
```
this branch is not needed as a no-op.

neolit123 replied:
i will leave this question to the api-machinery folks, but this makes sense to me:
> I think you would expect null, so I think I need to set out to zero in this branch.

neolit123 commented on yaml.go:
```diff
@@ -96,6 +96,61 @@ func JSONToYAML(j []byte) ([]byte, error) {
+			return nil
+		}
+		outV.Set(reflect.ValueOf(item))
+	} else {
```
handle `>0` here?

neolit123 replied:
adding this here `if len(outs) > 1 {` and removing the `if len(outs) == 0 {` branch.

chancez replied:
The `else` already is covering the ` > 1` in either case.

neolit123 replied:
if `if len(outs) == 0` is removed and if it remains a no-op, this branch has to handle `> 1`.

chancez commented:
I'm realizing that in the null/empty input cases I'm not overwriting `out` to be it's zero value/nil, and that seemed incorrect to me, but it seems similar to how the JSON unmarshaller behaves: https://play.golang.org/p/a5s0Z1md9L7

The only one im unsure of is what should be doing when:
- `out` is a non-zero value, aka: it could be a map with items in it.
- the input is an empty yaml stream: `---\n`

I've added a test for the other cases I tested in the playground link and I'm working on making the behavior pass the tests.

chancez commented:
I'm stuggling to figure out how to make https://github.com/kubernetes-sigs/yaml/pull/12/commits/b3c570ac403e5c01b56ab20c2ba655ac3ebd2ad5#diff-b9712b0e0447a2be5d319fb286dee2afR152 pass. Any tips would be good.

chancez commented:
Well, I've actually hit an even bigger issue in the reflection code that makes me realize the reflection here isn't correctly handling the `out` parameter's underlying type.  While working on benchmarking, I found this error:

```
panic: reflect.Set: value of type []interface {} is not assignable to type []*yaml.benchmarkTestItem
```

I understand what I need to do, just noting it here.

I'm also updating the benchmark to test `yaml.Unmarshal` and `internalYamlUnmarshal` directly so we can have both run as part of our tests to get a baseline.

neolit123 commented:
> panic: reflect.Set

i was going to ask if there is potential for panics in `Set`... :)

chancez commented:
I'm going to close this for now until I come up with a better solution/idea.
